File: modules/handlers/drive.py
"""
:project: telegram-onedrive
:author: L-ING
:copyright: (C) 2023 L-ING <hlf01@icloud.com>
:license: MIT, see LICENSE for more details.
"""

from telethon import events
import subprocess
from modules.client import tg_bot, onedrive
from modules.env import tg_user_name
from modules.utils import check_in_group, check_tg_login, cmd_parser
from modules.handlers.auth import od_auth
from modules.res import drive_res


@tg_bot.on(events.NewMessage(pattern="/drive", incoming=True, from_users=tg_user_name))
@check_in_group
@check_tg_login
async def drive_handler(event):
    cmd = cmd_parser(event)
    # /drive
    if len(cmd) == 1:
        message_str = ""

        try:
            users = onedrive.session.users
            current_user = onedrive.session.current_user
        except:
            await event.respond("No account found.")
            raise events.StopPropagation

        for i in range(len(users)):
            message_str += f"{i + 1}. {users[i]}\n"
        message_str += f"\nCurrent account is {current_user}\n"
        message_str += f"\n```/drive $index```\n To change current account."

        await event.respond(message_str)

    elif len(cmd) == 2:
        sub_cmd = cmd[1]
        # /drive add
        if sub_cmd == "add":
            auth_server = subprocess.Popen(("python", "server/auth_server.py"))
            async with tg_bot.conversation(event.chat_id) as conv:
                await od_auth(conv)
            auth_server.kill()

        # /drive logout
        elif sub_cmd == "logout":
            has_other_user = onedrive.logout()
            if has_other_user:
                current_user = onedrive.session.current_user
                logout_res = (
                    f"OneDrive logout successfully.\nCurrent account is {current_user}"
                )
            else:
                logout_res = "OneDrive logout successfully."
            await event.respond(logout_res)

        # /drive $index
        else:
            try:
                index = int(sub_cmd) - 1
            except ValueError:
                await event.reply("Account index should be integer.")
                raise events.StopPropagation

            try:
                users = onedrive.session.users
                old_user = onedrive.session.current_user
            except:
                await event.respond("No account found.")
                raise events.StopPropagation

            if index < len(users):
                new_user = onedrive.session.change_user(users[index])
                if old_user != new_user:
                    await event.respond(
                        f"Changed account from\n{old_user}\nto\n{new_user}"
                    )
                else:
                    await event.respond("Same account, nothing to change.")
            else:
                await event.reply("Account index out of range.")

    # /drive logout $index
    elif len(cmd) == 3:
        sub_cmd = cmd[1]
        if sub_cmd == "logout":
            try:
                index = int(cmd[2]) - 1
            except ValueError:
                await event.reply("Account index should be integer.")
                raise events.StopPropagation

            try:
                users = onedrive.session.users
            except:
                await event.respond("No account found.")
                raise events.StopPropagation

            if index < len(users):
                should_show_current_user = (
                    True if users[index] == onedrive.session.username else False
                )
                onedrive.logout(users[index])
                if should_show_current_user:
                    current_user = onedrive.session.current_user
                    logout_res = f"OneDrive logout successfully.\nCurrent account is {current_user}"
                else:
                    logout_res = "OneDrive logout successfully."
                await event.respond(logout_res)
            else:
                await event.reply("Account index out of range.")
        else:
            await event.reply(drive_res)
    raise events.StopPropagation

File: src/handlers/drive.rs
/*
:project: telegram-onedrive
:author: L-ING
:copyright: (C) 2024 L-ING <hlf01@icloud.com>
:license: MIT, see LICENSE for more details.
*/

use super::{
    docs::{format_help, format_unknown_command_help},
    utils::text::cmd_parser,
};
use crate::{
    auth_server, client::OneDriveClient, handlers::auth::authorize_onedrive,
    message::TelegramMessage, state::AppState,
};
use anyhow::{anyhow, Context, Result};
use grammers_client::InputMessage;
use proc_macros::{check_in_group, check_senders};

pub const PATTERN: &str = "/drive";

#[check_senders]
#[check_in_group]
pub async fn handler(message: TelegramMessage, state: AppState) -> Result<()> {
    let onedrive = &state.onedrive;

    let cmd = cmd_parser(message.text());

    if cmd.len() == 1 {
        // /drive
        show_drive(onedrive, message).await?;
    } else if cmd.len() == 2 {
        if cmd[1] == "add" {
            // /drive add
            add_drive(message, state.clone()).await?;
        } else if cmd[1] == "logout" {
            // /drive logout
            logout_current_drive(onedrive, message).await?;
        } else if cmd[1] == "help" {
            // /drive help
            message
                .respond(InputMessage::html(format_help(PATTERN)))
                .await
                .context("help")?;
        } else {
            // /drive $index
            let index = cmd[1]
                .parse::<usize>()
                .context("account index should be integer")?
                - 1;

            set_drive(onedrive, message, index).await?;
        }
    } else if cmd.len() == 3 {
        if cmd[1] == "logout" {
            // /drive logout $index
            let index = cmd[2]
                .parse::<usize>()
                .context("account index should be integer")?
                - 1;

            logout_drive(onedrive, message, index).await?;
        } else {
            return Err(anyhow!("sub command error")).context(format_unknown_command_help(PATTERN));
        }
    } else {
        return Err(anyhow!("command error")).context(format_unknown_command_help(PATTERN));
    }

    Ok(())
}

async fn show_drive(onedrive: &OneDriveClient, message: TelegramMessage) -> Result<()> {
    let usernames = onedrive.get_usernames().await?;
    if let Some(current_username) = onedrive.get_current_username().await? {
        if !usernames.is_empty() {
            let response = {
                let mut response = format!("Current account is {}", current_username);

                if usernames.len() > 1 {
                    response.insert(0, '\n');
                    for i in (1..=usernames.len()).rev() {
                        response.insert_str(0, &format!("{}. {}\n", i, usernames[i - 1]));
                    }
                }

                response
            };
            message.respond(response.as_str()).await.context(response)?;

            return Ok(());
        }
    }

    let response = "No account found.";
    message.respond(response).await.context(response)?;

    Ok(())
}

async fn add_drive(message: TelegramMessage, state: AppState) -> Result<()> {
    let (_, rx, _server_abort_handle) = auth_server::spawn().await?;
    authorize_onedrive(message, state, true, rx).await?;

    Ok(())
}

async fn logout_current_drive(onedrive: &OneDriveClient, message: TelegramMessage) -> Result<()> {
    let current_username = onedrive
        .get_current_username()
        .await?
        .ok_or_else(|| anyhow!("no onedrive account is logged in"))?;

    onedrive.logout(None).await?;

    let response = {
        let mut response = format!(
            "OneDrive account {} logged out successfully.",
            current_username
        );

        if let Some(current_username) = onedrive.get_current_username().await? {
            response.push_str(&format!("\n\nCurrent account is {}", current_username));
        }

        response
    };

    message.respond(response.as_str()).await.context(response)?;

    Ok(())
}

async fn set_drive(
    onedrive: &OneDriveClient,
    message: TelegramMessage,
    index: usize,
) -> Result<()> {
    let current_username = onedrive
        .get_current_username()
        .await?
        .ok_or_else(|| anyhow!("no onedrive account is logged in"))?;

    let usernames = onedrive.get_usernames().await?;

    let selected_username = usernames
        .get(index)
        .ok_or_else(|| anyhow!("account index out of range"))?;

    onedrive.change_account(selected_username).await?;

    if current_username == *selected_username {
        let response = "Same account, nothing to change.";
        message.respond(response).await.context(response)?;
    } else {
        let response = format!(
            "Changed account from\n{}\nto\n{}",
            current_username, selected_username
        );
        message.respond(response.as_str()).await.context(response)?;
    }

    Ok(())
}

async fn logout_drive(
    onedrive: &OneDriveClient,
    message: TelegramMessage,
    index: usize,
) -> Result<()> {
    let usernames = onedrive.get_usernames().await?;

    let selected_username = usernames
        .get(index)
        .ok_or_else(|| anyhow!("account index out of range"))?;

    onedrive.logout(Some(selected_username.clone())).await?;

    let response = {
        let mut response = format!(
            "OneDrive account {} logged out successfully.",
            selected_username
        );

        if let Some(current_username) = onedrive.get_current_username().await? {
            response.push_str(&format!("\n\nCurrent account is {}", current_username));
        }

        response
    };

    message.respond(response.as_str()).await.context(response)?;

    Ok(())
}
